KAFKA-4724: Clean up of state directories can possibly remove stores that are about to be used by another thread

dguy opened the pull request:
Delay the cleanup of state directories that are not locked and not owned by the current thread such that we only remove the directory if its last modified is < now - cleanupDelayMs.
This also helps to avoid a race between threads on the same instance, where during rebalance, one thread releases the lock on the state directory, and before another thread can take the lock, the cleanup runs and removes the data.

dguy committed:
delay cleanup of state directories

dguy commented:
@enothereska @guozhangwang @mjsax

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -203,8 +208,10 @@ public void cleanRemovedTasks() {
                     if (lock(id, 0)) {
-                        log.info("Deleting obsolete state directory {} for task {}", dirName, id);
-                        Utils.delete(taskDir);
+                        if (taskDir.lastModified() < time.milliseconds() - cleanupDelayMs) {
```
better: `if (time.milliseconds() > taskDir.lastModified() + cleanupDelayMs) { `
(easier to parse for humans, IMHO)

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -750,7 +750,7 @@ protected void maybeClean() {
         long now = time.milliseconds();
 
         if (now > lastCleanMs + cleanTimeMs) {
```
I guess, this condition can get removed now.

dguy replied:
we only want to run it infrequently, i.e., not every time around the runLoop. I think it should stay

mjsax replied:
Yes, but the same (or actually an improved) check is applied within `stateDirectory.cleanRemovedTasks(cleanTimeMs);` -- isn't it redundant?

dguy replied:
We only want to run the check on the interval. And then when it does run we only want to consider directories where the last mod is at least `cleanTimeMs` ago.

mjsax reviewed (commented):
Overall looks good to me.

dguy committed:
comments

guozhangwang commented:
@dguy LGTM overall. I have a couple of meta comments: I think the proposal is to take file's mtime into consideration, but I'm still wondering how much it can help with the race condition: if the other thread has not grabbed the lock, then this file will not be modified since its task was migrated out while store is flushed; if the other thread has already grabbed the lock, then we will not enter the critical path anyways, is that right?

One minor detailed comment: we need to update `StreamsConfig.STATE_CLEANUP_DELAY_MS_DOC` as well.

dguy commented:
@guozhangwang that is true, but in the usual case. The rebalance is happening reasonably quickly, so the chances of the directories being deleted is quite low. So it won't completely eliminate it, but generally it would be gone.

To completely eliminate we need to add some shared state in StreamThread so we know not to do the cleanup unless rebalancing has completed for all threads. I can do that - WDYT?

guozhangwang commented:
Discussed offline. I think it makes sense to me now. So the timeline would be sth. like:

```
t0: cleanup happened, lastCleanMs recorded;

t1: rebalance happend, store flushed;

t2: cleanup happend again, now > lastCleanMs + cleanTimeMs;
```

Before this patch, we are looking at `t2 - t0`, whereas in this patch we are looking at `t2 - t1`: it guarantees to give use the configured delay time after the rebalance no matter when the last cleanup function was triggered.

guozhangwang commented:
@dguy Could you rebase and update the config doc? Otherwise LGTM.

enothereska commented:
Will we open a JIRA for the un-usual case, just in case it happens? I'd prefer a more water-tight solution, but this might be sufficient for now. Thanks.

dguy committed:
fix merge conflicts

dguy commented:
@guozhangwang resolved the merge conflicts

dguy committed:
update doc and increase default value

dguy commented:
@guozhangwang i updated the config doc and i also changed the default cleanup interval to 10 minutes. 1 minute doesn't really seem that sensible to me.

@enothereska we can create a jira

guozhangwang reviewed (commented):
LGTM and merged to trunk.
